src/txn-wal/src/txn_read.rs

```diff
@@ -220,6 +220,36 @@ impl<T: Timestamp + Lattice + TotalOrder + Codec64> DataSnapshot<T> {
         data_since.snapshot_stats(as_of)
     }
 
+    /// See [ReadHandle::snapshot_stats].
+    pub fn snapshot_stats_from_leased<K, V, D>(
+        &self,
+        data_since: &ReadHandle<K, V, T, D>,
+    ) -> impl Future<Output = Result<SnapshotStats, Since<T>>> + Send + 'static
+    where
+        K: Debug + Codec + Ord,
+        V: Debug + Codec + Ord,
+        D: Ord + Semigroup + Codec64 + Send + Sync,
+    {
+        // This is used by the optimizer in planning to get cost statistics, so
+        // it can't use `unblock_read`. Instead use the "translated as_of"
+        // trick we invented but didn't end up using for read of the shard
+        // contents. The reason we didn't use it for that was because we'd have
+        // to deal with advancing timestamps of the updates we read, but the
+        // stats we return here don't have that issue.
+        //
+        // TODO: If we don't have a `latest_write`, then the `None` option to
+        // `snapshot_stats` is not quite correct because of pubsub races
```

Comment:
Sounds like we should return `Option<SnapshotStats>` here then? Why do we opt for returning possibly incorrect results instead of letting the caller deal with the unavailability of data?

Reply:
This is a straight copy of the existing method, so I didn't want to touch up how it behaves. But someone at some point might have to address these TODOs.